[resolve] centralize the standalone resolve and testing logic

rotty3000 opened the pull request.

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -103,94 +93,11 @@ private void testing(File runFile) throws MojoExecutionException, Exception, IOE
 
 			if (resolve) {
-				resolve(run);
+				run.resolve(failOnChanges);
```
The replaced `resolve` method did not modify the bndrun file. It just set the computed runbundles in the Run object: `run.setRunBundles(runBundles)`.  So we need to consider that here. Not sure what the right thing to do it, but I suspect that running test should not modify a source file. The export mojo modifies it as well. In the gradle plugin there is an explicit resolve task to modify the bndrun file. export only exports the -runbundles present.

bjhargrave replied:
I am thinking we should have a bnd-resolve-maven-plugin which can be used to run the resolve and modify the source file.

timothyjward replied:
A separate plugin focussed on resolving sounds good to me. It is definitely the maven way! +1

rotty3000 replied:
Does it matter that this would be a breaking change to the current usage?

bjhargrave replied:
So the change would be whether the resolve configuration flag results in mutating the bndrun source file. That seems fairly minor and can be documented as a breaking change. We can still honor the resolve flag, but it won't result in changing the bndrun source file.

bjhargrave replied:
So for this PR, I think you need to change the new Bndrun.resolve method such that you can either mutate the source bndrun file or not. And the method should probably return the new runbundles value, so that caller can use it if they don't request the mutation.

Then we can follow with a new PR to add the new bnd-resolve-maven-plugin.

rotty3000 replied:
Would this mean that the _new_ resolve plugin operation is not tied, by default, to any maven lifecycle and would need to be invoked manually?

bjhargrave replied:
> Would this mean that the new resolve plugin operation is not tied, by default, to any maven lifecycle and would need to be invoked manually?

Yes. Just like in Gradle and Bndtools. Running the resolve task is an explicit operation since it mutates a source file.

rotty3000 commented:
@bjhargrave can you look at this again? I've made the split and added a new bnd-resolver-maven-plugin which mutates the bndrun files but only does so explicitly as the plugin is not associated with any maven lifecycle.

Also, while I was there I added a table with all the configuration properties to each of the maven plugins.

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+				if (failOnChanges && !bemRunBundles.isEmpty()) {
+					String message = String.format("The runbundles have changed. Failing the build!\nWas: %s\nIs: %s",
+							originalRunbundlesString, runbundlesString);
+					throw new Exception(message);
```
Just throwing an exception seems wrong. You can all error since Bndrun is a Processor. Callers should check Bndrun for errors and then they can handle the error depending upon how they do it (gradle, maven, etc.)

rotty3000 replied:
Fixed!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+	public void test(File targetDir) throws Exception {
+		logger.info("Test " + this);
+		ProjectTester projectTester = getProjectTester();
+		File dir = new File(targetDir, "tmp");
```
Shouldn't we just use the supplied target dir as the directory?

rotty3000 replied:
I didn't write that logic so I'm not sure... although I don't see why not!

bjhargrave replied:
Seems like a good time to fix it. If we are going to not delete the test folder, it should at least be in the directory the caller specified.

rotty3000 replied:
Yup, fixed it!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+			logger.info("Warnings\n{}", Strings.join("\n", getWarnings()));
+		}
+
+		IO.delete(dir);
```
I don't think you want to always delete. Someone might want to look at the dir after a test failure occurs.

rotty3000 replied:
Fixed!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+			return;
+
+		logger.info("Errors\n{}", Strings.join("\n", getErrors()));
+		throw new Exception("Test had errors");
```
Also don't throw exception here. The errors should be in the Bndrun which the caller should check.

rotty3000 replied:
Fixed!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+		if (getErrors().isEmpty())
+			return;
+
+		logger.info("Errors\n{}", Strings.join("\n", getErrors()));
```
Leave the reporting to the caller.

rotty3000 replied:
Fixed!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+		projectTester.test();
+
+		if (!getWarnings().isEmpty()) {
+			logger.info("Warnings\n{}", Strings.join("\n", getWarnings()));
```
Leave the reporting to the caller.

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-export-maven-plugin/README.md:
```diff
@@ -26,3 +27,16 @@ a runnable jar into the targetDir. If not specified targetDir defaults to the pr
+|`resolve`              | Whether to resolve the `-runbundles` required for a valid runtime. _Defaults to `false`._|
+|`failOnChanges`        | Whether to fail the build if any change in the resolved `-runbundles` is discovered. _Defaults to `true`._|
+
+## Breaking Change (as of 3.4)
```
I would call this out in https://github.com/bndtools/bnd/wiki/Changes-in-3.4.0 rather than in this readme.

rotty3000 replied:
Ok!

bjhargrave commented on maven/bnd-export-maven-plugin/src/main/java/aQute/bnd/maven/export/plugin/ExportMojo.java:
```diff
@@ -87,78 +61,10 @@ private void export(File runFile) throws MojoExecutionException, Exception, IOEx
-				logger.info("{}:{}", Constants.RUNBUNDLES, bem.getDocumentChanges().get(Constants.RUNBUNDLES));
-				bem.saveChangesTo(doc);
-				IO.store(doc.get(), runFile);
+				run.resolve(failOnChanges, false);
```
After calling resolve, you should check for and report errors and warnings on the run object.
Also, since you pass false here, the resolve output it in the return value only which you did not collect. So don't you need to collect it and make sure it is used by the following export call? Otherwise the resolve options does nothing.

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-export-maven-plugin/src/main/java/aQute/bnd/maven/export/plugin/ExportMojo.java:
```diff
@@ -87,78 +61,10 @@ private void export(File runFile) throws MojoExecutionException, Exception, IOEx
-				IO.store(doc.get(), runFile);
+				run.resolve(failOnChanges, false);
 			}
+			run.export(targetDir);
```
Need to make sure resolve output, if any, from above is used. 

Also need to check run object for errors/warnings after export call and report.

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-resolver-maven-plugin/README.md:
```diff
@@ -0,0 +1,36 @@
+
+The `bnd-resolver-maven-plugin` is a bnd based plugin to resolve bundles from bndrun files.
+
+## What does the `bnd-resolver-maven-plugin` do?
```
Since this needs to be invoked "by hand", perhaps you should include a command line example of running it?

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-resolver-maven-plugin/pom.xml:
```diff
@@ -0,0 +1,112 @@
+        <dependency>
+            <groupId>commons-collections</groupId>
+            <artifactId>commons-collections</artifactId>
+            <version>3.2.2</version>
```
We don't need commons-collections any more since we it was only used to compare collections in the old code which called the resolver. Please remove references from all pom files (I think it is also in some other projects as leftovers.)

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-resolver-maven-plugin/pom.xml:
```diff
@@ -0,0 +1,112 @@
+                            <type>jar</type>
+                            <overWrite>false</overWrite>
+                            <outputDirectory>${project.build.outputDirectory}</outputDirectory>
+                            <destFileName>embedded-repo.jar</destFileName>
```
Does the bnd-resolve-maven-plugin need to embed the embedded-repo?

rotty3000 replied:
Hmm... testing...

bjhargrave commented on maven/bnd-resolver-maven-plugin/src/main/java/aQute/bnd/maven/resolver/plugin/ResolverMojo.java:
```diff
@@ -0,0 +1,63 @@
+			if (!run.isOk()) {
+				throw new MojoExecutionException("Initializing the workspace failed " + run.getErrors());
+			}
+			run.resolve(failOnChanges, true);
```
We need to check the run for errors/warnings and report them. And then fail the operation rather than the thrown Exception I objected to above.

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-resolver-maven-plugin/src/main/java/aQute/bnd/maven/resolver/plugin/ResolverMojo.java:
```diff
@@ -0,0 +1,63 @@
+
+	private void resolve(File runFile) throws MojoExecutionException, Exception, IOException {
+		if (!runFile.exists()) {
+			throw new MojoExecutionException("Could not find bnd run file " + runFile);
```
Since there can be multiple bndrun files, perhaps we should report errors/warnings for each file and then, after all files are attempted, throw a mojo exception if there was any error reported.

rotty3000 replied:
But we don't have a reporter until we have a runfile! Should we collect these in our own collection locally?

bjhargrave replied:
We can report the errors/warnings in the for loop over the runfiles. But this method should not throw the mojo exception. After the for loop, if any of the runfiles had an error, then throw the mojo exception.

rotty3000 replied:
This is the only thing I didn't do. I had enough changes already and wanted to get more of your thoughts on how to do this nicely. I can't see your goal here. I personally think failing fast is better than trying all the bndruns and only erroring at the end... but maybe I'm missing something.

bjhargrave replied:
For the same reason when you compile 20 java source files, the compiler gives you errors and warnings on all the source files before ending compilation. If the compiler gave up at the first source file, it would take many compiler iterations to solve all the errors. This is build time tool and should help the developer as much as possible.

Failfast is good when you need to prevent data corruption. I don't think that is the case here :-)

rotty3000 replied:
What's the message of the mojoexceptions then? are you suggesting that we give nothing as message there? which error should I use as the final one? Do I show all of them as the message?

bjhargrave replied:
I imagine it would be something like: "XX errors found" where XX is the total number of errors. This should be similar to the mojo exception from the compiler plugin. All the errors are error logged, so that are already displayed, and the moj exception is just a summary of the errors. The gradle plugin does this sort of display errors and then throw exception with summary text: https://github.com/bndtools/bnd/blob/master/biz.aQute.bnd.gradle/src/aQute/bnd/gradle/BndPlugin.groovy#L645.

rotty3000 replied:
I did my best to do what I think you want... See updates.

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -103,94 +93,11 @@ private void testing(File runFile) throws MojoExecutionException, Exception, IOE
 
 			if (resolve) {
-				resolve(run);
+				run.resolve(failOnChanges, false);
```
We have the same issue here since resolve is called with false. How is the resolve output used by the following test execution?

Also need to check for and report errors/warnings.

rotty3000 replied:
Fixed!

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -103,94 +93,11 @@ private void testing(File runFile) throws MojoExecutionException, Exception, IOE
 			}
 			run.setProperty("-runbundles.tester", "biz.aQute.tester");
-			testing(run);
+			run.test(targetDir);
```
Need to check for errors/warnings and report.

rotty3000 replied:
Fixed!

bjhargrave commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,146 @@
+					IO.store(doc.get(), runFile);
+				}
+			}
+			return headerClauseListFormatter.convert(bem.getRunBundles());
```
This return value is never used. And the type is not useful. You need to return some thing that can be used by the caller to call `setRunBundles` on this run object. So when someone calls resolve with `writeOnChanges` = false, they can use the return value to call `setRunBundles` before they then call `export` or `test` (assuming those methods use the set runbundles value.

rotty3000 replied:
The setRunBundles method is pretty useless actually because I have to apply the settings to the project inorder to generate a `Collection<Container>` which is pointless because that already sets if before the caller would decide if they wanted to or not! The only way to let them optionally set the runbundles is to use `run.setProperty(Constants.RUNBUNDLES, resolveReturnValue);`

bjhargrave replied:
There must be a way to set the runbundles computed by  the resolver into the Run object so that test and export will use that value instead of the value in the bndrun file (which may not even be present).

rotty3000 replied:
yes... using `setProperty`!

bjhargrave commented:
Also, I think some more tests are probably needed for bnd-test-maven-plugin and bnd-export-maven-plugin to make sure the bndrun file is not altered.

rotty3000 commented:
Lots of updates. @bjhargrave can you take a look again? 

Two things I didn't do yet were:
- change the logic of the bndrun loops
- add tests to assert testing and exporting don't change the bndrun file

bjhargrave commented on maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/verify.groovy:
```diff
@@ -0,0 +1,30 @@
+
+// Get the -runbundles.
+def bemRunBundles = bem.getRunBundles()
+assert null != bemRunBundles
```
This is groovy, the assert uses groovy truthiness. Non-null is true and null is false. So just `assert bemRunBundles()`

bjhargrave commented on maven/bnd-resolver-maven-plugin/src/test/resources/integration-test/test/verify.groovy:
```diff
@@ -0,0 +1,30 @@
+
+StringBuilder sb = new StringBuilder();
+bemRunBundles.get(0).formatTo(sb);
+assert sb.toString().equals("org.apache.felix.eventadmin;version='[1.4.6,1.4.7)'");
```
This is groovy, so == means equals. So just `assert sb.toString() == "org.apache.felix.eventadmin;version='[1.4.6,1.4.7)'"`

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -44,7 +34,7 @@
 	private List<File>	bndruns;
 
-	@Parameter(defaultValue = "${project.build.directory}", readonly = true)
+	@Parameter(defaultValue = "${project.build.directory}/tmp", readonly = true)
```
As I discussed above, we need 2 args: cwd to run tests in and reportsDir. I am sure `target/tmp` is a bad choice. We need a basedir to use for cwd of each bndrun file and the use the bndrun file name for a subfolder.

"${project.build.directory}/test/${bndrun}"  for cwd and "${project.build.directory}/test-reports/${bndrun}" for reports dir. So the parameters to this plugin are "${project.build.directory}/test" and "${project.build.directory}/test-reports"

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -85,112 +77,48 @@ public void execute() throws MojoExecutionException, MojoFailureException {
-					logger.warn("The runbundles have changed:");
-					run.setRunBundles(runBundles);
-				}
+			run.test(targetDir);
```
Need to make cwd and reportsDir for each bndrun file and pass both as args.

rotty3000 commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,162 @@
+		targetDir.mkdirs();
+
+		projectTester.setCwd(targetDir);
+		projectTester.setReportDir(targetDir);
```
So should there be a delete operation here?

rotty3000 commented on maven/bnd-export-maven-plugin/src/main/java/aQute/bnd/maven/export/plugin/ExportMojo.java:
```diff
@@ -70,95 +50,46 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 			run.check();
 			if (!run.isOk()) {
-				throw new MojoExecutionException("Initializing the workspace failed " + run.getErrors());
+				logger.error("Initializing the workspace failed " + run.getErrors());
```
I think we need to warn _always_ and only report error in the `!run.isOk()` case right which already checks for errors but also checks if `-failok` flag and we already decided to return at that point, right?

rotty3000 commented:
Updated! I believe I made all the requested changes.

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -85,112 +81,42 @@ public void execute() throws MojoExecutionException, MojoFailureException {
 				}
+				run.setProperty(Constants.RUNBUNDLES, runBundles);
 			}
+			run.setProperty("-runbundles.tester", "biz.aQute.tester");
```
This should be `Constants.TESTER` not `"-runbundles.tester"`. I grepped the code and the only place `-runbundles.tester` is used it right here. See https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/src/aQute/bnd/build/Project.java#L2486-L2491. The default for `Constants.TESTER` is `"biz.aQute.tester"` which is why this no-op does not hurt. So I think you could safely delete this line which would allow the bndrun file to specify its own tester.

bjhargrave replied:
Wait, I get it. This is adding another bundle to -runbundles as a merged property.

bjhargrave replied:
But it is still not needed I think because biz.aQute.tester adds itself to runbundles anyway: https://github.com/bndtools/bnd/blob/master/biz.aQute.tester/src/aQute/tester/plugin/ProjectTesterImpl.java#L57

bjhargrave commented on maven/bnd-testing-maven-plugin/src/main/java/aQute/bnd/maven/testing/plugin/TestingMojo.java:
```diff
@@ -85,112 +81,42 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+				run.setProperty(Constants.RUNBUNDLES, runBundles);
 			}
+			run.setProperty("-runbundles.tester", "biz.aQute.tester");
+			String bndrun = runFile.getName();
```
Shouldn't this do the getnamepart thing to remove the '.bndrun' suffix?

rotty3000 commented on maven/bnd-export-maven-plugin/src/main/java/aQute/bnd/maven/export/plugin/ExportMojo.java:
```diff
@@ -59,8 +41,7 @@
 
-	private Converter<String,Collection< ? extends HeaderClause>>	headerClauseListFormatter	= new CollectionFormatter<HeaderClause>(
-			",\\\n\t", new HeaderClauseFormatter(), null);
+	private AtomicInteger		errors	= new AtomicInteger();
```
because the int would be pass by value?

rotty3000 commented on biz.aQute.resolve/src/biz/aQute/resolve/Bndrun.java:
```diff
@@ -0,0 +1,188 @@
+		return projectTester.test();
+	}
+
+	public void error(AtomicInteger count, String message, Object... args) {
```
so I have to replicate the same methods 3 times? seems bad.

bjhargrave replied:
Well this is a maven plugin issue. The maven plugins need shared code. How do maven plugins share code?

Just because it is hard to share code across maven plugins does not mean you push the code up into bnd.

bjhargrave replied:
You will see that bnd-maven-plugin needs its own code to check errors and warnings. You may want to consider using that to play nice with m2e:  https://github.com/bndtools/bnd/blob/master/maven/bnd-maven-plugin/src/main/java/aQute/bnd/maven/plugin/BndMavenPlugin.java#L328-L373

rotty3000 replied:
Fixed!

rotty3000 replied:
... well fixed in the simple way.. not in that last BndMavenPlugin/m2e way...

rotty3000 commented:
Updated again!

rotty3000 commented:
Updated! gotta add some no-write-on-bndrun tests for bnd-exporter and bnd-testing

rotty3000 commented:
ok, tests added!

gamerson commented on maven/bnd-resolver-maven-plugin/src/main/java/aQute/bnd/maven/resolver/plugin/ResolverMojo.java:
```diff
@@ -0,0 +1,84 @@
+package aQute.bnd.maven.resolver.plugin;
```
Hey @rotty3000  do we need some m2e lifecycle-mappings for this new mojo?

gamerson replied:
Actually do we expect this mojo to mainly be on projects with packaging type 'pom'?  If so then lifecycle mappings are not needed.

bjhargrave replied:
This plugin will not run during a normal build. It is mostly for command line use to mutate a bndrun file's -runbundles. So I don't think we need an m2e mapping.

gamerson commented:
Ok, sounds good, I'll get Ray to provide me with some sample projects that
I can import into Eclipse/m2e and verify that it doesn't get fussy about
lifecycle mappings that it doesn't understand.  It should since it ignores
mappings for all pom packaging modules.  But maybe we can also provide some
'jar' packaging module samples that include this plugin to make sure all
mappings are safely ignored (no annoying red X's)

On Fri, Feb 17, 2017 at 7:20 AM, BJ Hargrave <notifications@github.com>
wrote:

> *@bjhargrave* commented on this pull request.
> ------------------------------
>
> In maven/bnd-resolver-maven-plugin/src/main/java/aQute/
> bnd/maven/resolver/plugin/ResolverMojo.java
> <https://github.com/bndtools/bnd/pull/1876#discussion_r101754422>:
>
> > @@ -0,0 +1,84 @@
> +package aQute.bnd.maven.resolver.plugin;
>
> This plugin will not run during a normal build. It is mostly for command
> line use to mutate a bndrun file's -runbundles. So I don't think we need an
> m2e mapping.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/bndtools/bnd/pull/1876#discussion_r101754422>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAkVFRLuH6pONcrIJWMqXo92r4pAWdk2ks5rdZ6igaJpZM4L2x5X>
> .
>



-- 
Greg Amerson
Liferay Developer Tools
Liferay, Inc. www.liferay.com

bjhargrave commented:
LGTM. Once merged, I'll change the gradle plugin over to use this new support and also add tasks to the non-workspace gradle plugin.
